Steam project pull request

DvdHalliday opened the pull request:
This was all done according to instructions

DvdHalliday committed:
Part One

Added the classes Videogame and Date with its respective header and cpp files.

DvdHalliday committed:
Part One Complement

Forgot to add getters for the private variables which we will eventually make use of.

DvdHalliday committed:
Part Two

Added the files in charge of creating the Category class.

DvdHalliday committed:
Part Two Complement

Changed the CanAddGame name function to IsFull

DvdHalliday committed:
Part Three

Added the files needed to create de CategoryContainer class.

DvdHalliday committed:
Part Three Complement

Added the other condition on the for statement so it wouldn't have any trouble in the cases where CurrentNumberOfCategories is exactly the same as the MaxCategoriesAmount so we avoid calling an index out of bounds

DvdHalliday committed:
Part Four

Added the Steam header file and the file that contains the main method

DvdHalliday committed:
Part Five

Added the Main Menu and the Add Game Menu with its respective functions. I also added some other supporting functions like Input validators and such.

DvdHalliday committed:
Part Six WorkInProgress

It essentially works, however there is something wrong when creating a category as it isn't creating it with the given name. In this commit I add the functionalities for the Category Manager Menu.

DvdHalliday committed:
Part Seven WorkInProgress

Added the functionalities to display all videogames. It doesn't quite work with the games that are inside a category the user creates. I think it has something to do with the problems I got in part six when dealing with category constructors.

DvdHalliday committed:
Part Six and Seven Complement

Fixed several problems, added a function in categorycontainer since the original format wasn't allowing the modification of a Category retrieved by a GetCategory.
Modified another problem in the function OpenRemoveCategory since there was a wrong index. I also changed a behaviour in the Game displayer, it was returning the games from the uncategorized category where it should've been returning a specific category.

DvdHalliday committed:
Adding const correctness and comment separators

This was done throughout all the files that hadn't been updated according to these standards.

DvdHalliday committed:
Uploading other files

.sln and the .vcxproj files

simonkytr commented on c++ training/SteamProject/Videogame/Date.cpp:
```diff
@@ -0,0 +1,40 @@
+std::string FDate::ToFormattedString() const
+{
+	const std::string ReturnString = std::to_string(GetDay()) + "/" + std::to_string(GetMonth()) + "/" + std::to_string(GetYear());
+	return ReturnString;
```
you should return the statement directly 

```suggestion
	return std::to_string(GetDay()) + "/" + std::to_string(GetMonth()) + "/" + std::to_string(GetYear());
```

simonkytr commented on c++ training/SteamProject/Videogame/Date.cpp:
```diff
@@ -0,0 +1,40 @@
+FDate::FDate(int Day, int Month, int Year)
+{
+	SetDate(Day, Month, Year);
+}
```
you could create the constructor to initialize the object directly instead to initialize it indirectly with a function.

simonkytr commented on c++ training/SteamProject/Videogame/Category.cpp:
```diff
@@ -0,0 +1,40 @@
+	{
+		return Games[Index];
+	}
+	return FVideogame();
```
inside the if you are already returning an FVideoGame object because Games[Index] is an array of FVideoGames

DvdHalliday replied:
Yup, however we should always return something when dealing with a function that has a return type. That simply returns a new default object in case the condition on the if statement isn't reached.

simonkytr commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+	enum class MainMenuCommand { AddGameMenu = 1, CategoryManagerMenu, GameDisplayerMenu, Exit };
+	MainMenuCommand UserChoice;
+
+	float TemporalUserChoice;
```
what do you mean about `TemporalUserChoice`?
what do you think about `UserChoiceMenu` or something seems like?

DvdHalliday replied:
it's temporal since it is likely to change, I'm dealing with floats because I intend to avoid cases such as 2.1 or other floats or doubles being recognized as the integer 2. So at this point it is just a possibility that variable's gonna be the one the app actually decides to run with. Hence the Temporal in the name.

JDAlzate replied:
Temporal could be a good enough name. It may be better to find a better distinction, such as `UserInput` which would differ from `UserChoice`, since to me Temporal sounds as if the user would change their mind, not that the choice will be adjusted, but that may be just me.

However, there is a better solution here, which would also affect the previous comment about the method name: why do we need to have the first std::cin outside of the validation method?

If the method was named something like: `GetValidInput(...)` and the function itself could handle getting the user input. That way, the variable TemporalUserChoice will never exist in the context of OpenMainMenu and so it could never be confused with UserChoice and we don't have to worry about differentiating it. 

This would be nice since the function always asks for an input, so why not instead get the input from within the function? Parameters are useful when the value can't be gotten internally, but in this case, it can. What do you think?

This would simply much of the code in  your review :)

simonkytr commented on c++ training/SteamProject/Videogame/Steam.h:
```diff
@@ -0,0 +1,47 @@
+
+	bool IsInteger(const float Float) const;
+	bool IsInRange(const float Value, const int LowerBound, const int UpperBound) const;
+	int ValidateInputAndReturnValidatedInput(float FirstInput, const std::string Message, const int LowerBound, const int UpperBound);
```
about this name... we have to be the most descriptive possible but is it necessary that its function have a name so large?

DvdHalliday replied:
I mean, I also hesitated when I named it, the ValidateInput name could tell me it's trying to validate the input, but it isn't quite clear I want a validated input to be returned. That could, however, be dealt with pointers, though we cannot really use them yet, so I'm a bit confused here on how to proceed.

JDAlzate replied:
You're right that this can be a tricky method to name. The important part is for the return type to be clear and I do believe that in this particular case, the int being returned, from a validation function, is assumed to be the validated input.

You could do a couple things:
1. Commenting is great here. A simple comment saying that it returns the validated input goes a long way.
2. I think considering `ValidateInput` vs `GetValidatedInput`, the second option would be clearer since you are certainly getting the validated input. Sure, you're requesting the user for additional input, but I believe that would clasify as _getting_ the validated input. Certainly, validation is not a simple getter... something must be going on under the hood.

But I do believe that the current method name is overly specific. What do you think?

simonkytr commented on c++ training/SteamProject/Videogame/Steam.h:
```diff
@@ -0,0 +1,47 @@
+	bool IsInteger(const float Float) const;
+	bool IsInRange(const float Value, const int LowerBound, const int UpperBound) const;
+	int ValidateInputAndReturnValidatedInput(float FirstInput, const std::string Message, const int LowerBound, const int UpperBound);
+	std::string ValidateInputAndReturnValidatedInput(std::string StringInput, const std::string Message);
```
the same as the previous comment

simonkytr commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+	}
+
+	std::cout << "Please enter a name for the new category: ";
+	std::string TemporaryCategoryName;
```
you should condense the name `TemporaryCategoryName` to just `NewCategoryName` or `CategoryName`

DvdHalliday replied:
Same as before, it might not be the actual NewCategoryName, but just a first attempt given to us by the user.

DvdHalliday replied:
However I think I'm gonna' change the name to TemporalCategoryName to follow the example of the ones I created before.

simonkytr commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+	std::string CategoryManagerMessage = "What would you rather do next? \n1 - Add a new category\n2 - Remove a category\n3 - Go back to main menu\nTELL ME : ";
+	std::cout << CategoryManagerMessage;
+
+	enum class CategoryManagerCommand { CreateCategory = 1, RemoveCategory, MainMenu };
```
according to the coding standards, the enums have to be named with the prefix `E`

simonkytr commented on c++ training/SteamProject/Videogame/Steam.h:
```diff
@@ -0,0 +1,47 @@
+
+	int NumberOfGames;
+
+	enum class MenuCommand
```
according to the coding standards, the enums have to be named with the prefix `E`

simonkytr commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+	std::string AskChooseCategoryMessage = "Would you like to choose a category for your new game? \n1 - Yes\n2 - No\nTELL ME : ";
+	std::cout << AskChooseCategoryMessage;
+
+	enum class QuestionCategoryCommand { Yes = 1, No };
```
according to the coding standards, the enums have to be named with the prefix `E`

DvdHalliday committed:
Diego's code review commit

Corrected the way I was initializing the FDate objects and the way I returned the formatted string.
Added the E prefixes on all enums in the project and changed the Temporary in a variable to Temporal to maintain consistency throughout the project.

JDAlzate commented on c++ training/SteamProject/Videogame/Date.h:
```diff
@@ -0,0 +1,21 @@
+#include<iostream>
+#include<string>
+
+class FDate {
```
This class is looking nice! :) Good that you decided to include it. 

Be careful with the braces though, the convention asks for braces to be on a new line.

JDAlzate commented on c++ training/SteamProject/Videogame/Date.cpp:
```diff
@@ -0,0 +1,37 @@
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+std::string FDate::ToFormattedString() const
+{
+	return std::to_string(GetDay()) + "/" + std::to_string(GetMonth()) + "/" + std::to_string(GetYear());;
```
Double semicolon here.

JDAlzate commented on c++ training/SteamProject/Videogame/Videogame.h:
```diff
@@ -0,0 +1,20 @@
+#include <iostream>
+#include "Date.h"
+
+class FVideogame {
```
The brace style goes for the rest of the review as well. Seems like you did this for all classes.

JDAlzate commented on c++ training/SteamProject/Videogame/Videogame.cpp:
```diff
@@ -0,0 +1,32 @@
+#include "Videogame.h"
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+FVideogame::FVideogame() : Name(""), Studio(""), Date(FDate())
```
Keep in mind that classes with default constructors don't need to be explicitly initialized, as their constructors are called upon definition. So, there's really no need to explicitly initialize any of the member variables in this class :)

Please fix this for the rest of your review

JDAlzate commented on c++ training/SteamProject/Videogame/Videogame.cpp:
```diff
@@ -0,0 +1,32 @@
+{
+}
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+FVideogame::FVideogame(std::string NewName, std::string NewStudio, FDate NewDate) : Name(NewName), Studio(NewStudio), Date(NewDate)
```
All of these parameters could be const

JDAlzate commented on c++ training/SteamProject/Videogame/Videogame.cpp:
```diff
@@ -0,0 +1,32 @@
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+bool FVideogame::IsValid() const
+{
+	const bool bIsValidDate = Date.GetDay() > 0 && Date.GetDay() < 32 &&
```
This is great, but this falls into the responsibility of the date class, don't you think? Instead of having to manually check date validity here, what if the `FDate` class itself had an `IsValid()` public method? That way, you could check validity whenever you want, without having to duplicate this code.

JDAlzate commented on c++ training/SteamProject/Videogame/Videogame.cpp:
```diff
@@ -0,0 +1,32 @@
+	const bool bIsValidDate = Date.GetDay() > 0 && Date.GetDay() < 32 &&
+		Date.GetMonth() > 0 && Date.GetMonth() < 13 &&
+		Date.GetYear() > 1900 && Date.GetDay() < 2023;
+	return (Name != "") && (Studio != "") && bIsValidDate;
```
Though this is working great, the lines Name != "" and Studio != "" are actually creating a new empty string literal to check against it. That's not too bad, but it would be better to use built-in functions rather than comparing the string to other strings. In this case, the std::string class has a function called empty() that you could use.

Please fix this for the rest of your review
See https://cplusplus.com/reference/string/string/ for a reference to std::string and its available methods.

JDAlzate commented on c++ training/SteamProject/Videogame/Category.cpp:
```diff
@@ -0,0 +1,40 @@
+{
+}
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+FCategory::FCategory(std::string NewName) : Name(NewName), CurrentNumberOfGames(0)
```
Remember that when parameters do not change in the body of the function, they can be const. Please fix this for the rest of your review.

JDAlzate commented on c++ training/SteamProject/Videogame/Category.cpp:
```diff
@@ -0,0 +1,40 @@
+void FCategory::AddGame(FVideogame Videogame)
+{
+	// This adds the game in the current CurrentNumberOfGames position and after that increases its value adding 1
+	Games[CurrentNumberOfGames++] = Videogame;
```
What if I try to add a game when the list is full?

JDAlzate commented on c++ training/SteamProject/Videogame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,59 @@
+		Categories[i] = Categories[i + 1];
+	}
+	CurrentNumberOfCategories--;
+	return true;
```
You're returning true even if the category wasn't removed. What if I enter the index 50164?

JDAlzate commented on c++ training/SteamProject/Videogame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,59 @@
+	{
+		return false;
+	}
+	for (int i = Index; i < CurrentNumberOfCategories && i < MaxCategoriesAmount-1; i++)
```
I think your condition is overly complicated. Think about it, how would you make your loop condition use just the CategoryCount without the need for MaxCategoryCount?

JDAlzate commented on c++ training/SteamProject/Videogame/CategoryContainer.cpp:
```diff
@@ -0,0 +1,59 @@
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+void FCategoryContainer::AddGameToCategory(const int CategoryIndex, const FVideogame Game)
+{
+	Categories[CategoryIndex].AddGame(Game);
```
what if I tried adding a game in the category of index 66? It's good to account for user errors in our functions.

JDAlzate commented on c++ training/SteamProject/Videogame/Source.cpp:
```diff
@@ -0,0 +1,7 @@
+#include "Steam.h"
+
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+int main() {
```
Main should also use the brace convention.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+		break;
+
+	default:
+		std::cout << "MainMenu Error, please contact support";
```
Adding a break to the default case is useful if in the future some case was added below the default one.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+
+	switch (UserChoice)
+	{
+	case EMainMenuCommand::AddGameMenu:
```
Since you're validating the input, you're absolutely certain that the input is within the range of the enum, so why not assign it directly? That way, you wouldn't even need an additional variable for UserChoice.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+
+	if (!CategoryContainer.IsEmpty())
+	{
+		CategoryIndex = ChooseCategory();
```
This is a bit overcomplicated and dangerous (in the sense that it is not accounting for user error). Your `ChooseCategory` function is always assuming that you only call it if the category container is empty... but couldn't that be part of the ChooseCategory function itself?

Instead of 
```
if (!AskChooseCategory())
{
    return -1;
}
```
You could have
```
if (CategoryContainer.IsEmpty() || !AskChooseCategory())
{
    return -1;
}
```
and this way, you don't have to concern yourself with validating the CategoryContainer before calling ChooseCategory() and ChooseCategory() wouldn't be possibly used wrong and cause a weird bug.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+		std::cout << "There are currently no games to display. Please type Enter to return to the main menu. ";
+
+		std::cin.ignore(500, '\n');
+		if (std::cin.peek() == '\n')
```
What if I don't press enter, but rather some other key?

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+		std::cin.ignore(500, '\n');
+		if (std::cin.peek() == '\n')
+		{
+			OpenMainMenu();
```
In this case, you would be opening the main menu without having the active choice be the main menu.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+		}
+	}
+
+	if (!CategoryContainer.IsEmpty())
```
This check is unnecessary here. Your loop won't run anything if the category container is empty, so this is just decreasing legibility by adding an extra nested block of code.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+		std::cout << "There is no available space right now to create another category, please type Enter to return to the previous menu";
+
+		std::cin.ignore(500, '\n');
+		if (std::cin.peek() == '\n')
```
This is repeated a lot in your code, but you're never accounting for what would happen if you pressed something else.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+	{
+		OpenCategoryManagerMenu();
+	}
+	return;
```
There's no need to return here, you are already at the end of the function.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+		OpenCategoryManagerMenu();
+	}
+	return;
+}
```
Same here.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+	return;
+}
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+int FSteam::ChooseCategory()
```
This method could be const. (It couldn't currently, but your validation functions could also be const. If those are const, this one can be as well)

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+	return ValidateInputAndReturnValidatedInput(TemporalUserChoice, "TELL ME A VALID OPTION: ", 1, CategoryContainer.GetCurrentNumberOfCategories()) - 1;
+}
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+bool FSteam::AskChooseCategory()
```
Same here. This could be const.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+	return Value >= LowerBound && Value <= UpperBound;
+}
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+int FSteam::ValidateInputAndReturnValidatedInput(float Input, std::string Message, int LowerBound, int UpperBound)
```
This and the next method could be const.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+	return StringInput;
+}
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+int FSteam::GetValidDay(float TemporalDay, int Month, int Year)
```
This one could also be const after making your validation methods constant. This is an example of the issues with having even just one method not follow const-correctness: you're forced to have more non-const methods.

JDAlzate commented on c++ training/SteamProject/Videogame/Steam.cpp:
```diff
@@ -0,0 +1,434 @@
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+void FSteam::RunSteam()
+{
+	while (true)
```
This is perfectly fine, but better reserved for cases where it's hard to specify the condition. In this case it's easy (when active command isn't exit) so it's better to be explicit about it.

JDAlzate reviewed (changes requested):
Review done! Good job with the project! :)

DvdHalliday committed:
JD's code review commit

I tried to get done all the changes requested in that review
